server/src/main/java/org/elasticsearch/common/time/JavaDateFormatter.java

```diff
@@ -122,30 +114,65 @@ public TemporalAccessor parse(String input) {
         }
 
         try {
-            return parser.parse(input);
+            return doParse(input);
         } catch (DateTimeParseException e) {
             throw new IllegalArgumentException("failed to parse date field [" + input + "] with format [" + format + "]", e);
         }
     }
 
+    private TemporalAccessor doParse(String input) {
+        if (parsers.size() > 1) {
+            for (DateTimeFormatter formatter : parsers) {
```

Comment:
We should not do this. It is exactly what we got rid of that caused massive performance degradation when using java time vs joda. I think we should fail when parsing the format if there is ambiguity.

Reply:
I can't exactly agree if there really is ambiguity. If there is a composite pattern where the first pattern is a prefix of the second pattern (as in the example from the issue) then only the second one is really matching the date format.
This is the way it used to work in joda and we could support this. 

It would be hard to explain in our doc why we forbid patterns like the one from the issue. Java-time javadoc does not mention this limitation.

There is obviously a performance drop (~14%), not sure if this is still acceptable. Will run Rally benchmarks later.

@polyfractal @Mpdreamz what are your views on this?

Reply:
No real opinion on implementation details, but I do think it's a huge breaking change if we can't support the old multi-format behavior. Even ignoring what we've supported in the past, `yyyy-MM-dd'T'HH:mm:ss||yyyy-MM-dd'T'HH:mm:ss.SSS` doesn't seem ambiguous/wrong/incorrect if you have both formats in your data.  

And situations like `yyyy-MM-dd'T'HH:mm:ss||date_optional_time` might be technically redundant but due to our naming scheme it's not clear to the user what's going on.

And I'm sure there are thousands of legacy templates with patterns like those that will break on upgrade.

Wasn't the original performance issues due to throwing and catching exceptions as control-flow, rather than trying multiple parsers?

Reply:
I agree with @polyfractal: The root cause of the performance regression (https://github.com/elastic/elasticsearch/issues/36602) was not that multiple parsers have been applied but rather that each of them has thrown an `IllegalArgumentException` on mismatch. I think dropping support for multiple date patterns is not an option.

Reply:
I have refactored this method to use `DateTimeFormatter.format.parseObject`. The implementation in `DateTimeFormatter::ClassicFormat` is returning null when parsing failed. This has the same complexity as the approach currently in master. `DateTimeFormatterBuilder:2349` is also iterating within enclosed optionals in a loop (but has this downside of returning when input is fully consumed)
see the results of Rally and microbenchmark on this PR #40364. These look very promising to me.